rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,297 @@
+    List<ResolvedExpression> indexKeyFilters = dataFilters.stream().filter(expr -> ExpressionUtils.isEqualsLitExpr(expr, indexKeyFields)).collect(Collectors.toList());
+    if (!ExpressionUtils.isFilteringByAllFields(indexKeyFilters, indexKeyFields)) {
+      return PrimaryKeyPruners.BUCKET_ID_NO_PRUNING;
+    }
+    return PrimaryKeyPruners.getBucketId(indexKeyFilters, conf);
+  }
+```
+**Step 2**: Do partition pruning and get all files in given partitions
+**Step 3**: do bucket pruning for all files from step2
+```java
+  /**
+   * Returns all the file statuses under the table base path.
+   */
+  public List<StoragePathInfo> getFilesInPartitions() {
+    ...
+    // Partition pruning
+    String[] partitions =
+        getOrBuildPartitionPaths().stream().map(p -> fullPartitionPath(path, p)).toArray(String[]::new);
+    if (partitions.length < 1) {
+      return Collections.emptyList();
+    }
+    List<StoragePathInfo> allFiles = ...
+    
+    // bucket pruning
+    if (this.dataBucket >= 0) {
+      String bucketIdStr = BucketIdentifier.bucketIdStr(this.dataBucket);
+      List<StoragePathInfo> filesAfterBucketPruning = allFiles.stream()
+          .filter(fileInfo -> fileInfo.getPath().getName().contains(bucketIdStr))
+          .collect(Collectors.toList());
+      logPruningMsg(allFiles.size(), filesAfterBucketPruning.size(), "bucket pruning");
+      allFiles = filesAfterBucketPruning;
+    }
+    ...
+  }
+
+```
+
+### Spark Write/Read Using Simple Bucket Index
+The read-write process of Spark based on Bucket Index is also similar.
+- Use `HoodieSimpleBucketIndex` to tag location.
+- Use `SparkBucketIndexPartitioner` to packs incoming records to be inserted into buckets (1 bucket = 1 RDD partition).
+- Use `BucketIndexSupport` to Bucket Index pruning during reading.
+
+## Design
+### Config
+Add a new config named `hoodie.bucket.index.partition.expressions` default null. Users can specify the bucket numbers for different
+partitions by configuring a JSON expression. For example
+```json
+{
+    "expressions": [
+        {
+            "expression": "11-11",
+            "bucketNumber": 10,
+            "rule": "regex"
+        },
+        {
+            "expression": "01-01",
+            "bucketNumber": 20,
+            "rule": "regex"
+        },
+        {
+            "expression": "dt>2025-01-01",
+            "bucketNumber": 20,
+            "rule": "range"
+        }
+    ]
+}
+```
+for partitions match different rule will get and set corresponding bucket number.
+
+We can determine whether the user is currently using the partition-level bucket index based on the value of
+`hoodie.bucket.index.partition.expressions`. If it is null, the processing behavior will be exactly the same as the current logic.
+The advantage of this approach is that it can be fully compatible with the current design of the table-level bucket index,
+enabling a seamless migration for users without their awareness.
+
+### Hashing Metadata
```

Comment:
Can we elaborate the functionality first.

Reply:
Sure thing. Added overview part.